crates/bin/node/src/grpc_service.rs

```diff
@@ -259,6 +273,55 @@ impl Node for GrpcState {
             })
             .collect::<Result<Vec<_>, _>>()?;
 
+        let mut conn = match self.pg_pool.acquire().await {
+            Ok(conn) => conn,
+            Err(e) => return Err(Status::internal(e.to_string())),
+        };
+
+        for input in &inputs {
+            let keyset_info = match db_node::keyset::get_keyset::<String>(&mut conn, &input.keyset_id).await {
+                Ok(info) => info,
+                Err(e) => return Err(Status::internal(e.to_string())),
+            };
+            
+            let max_order_u8 = keyset_info.max_order();
+            let max_value = if max_order_u8 >= 64 {
+                u64::MAX
+            } else {
+                (1u64 << max_order_u8) - 1
+            };
+            
+            if u64::from(input.amount) > max_value {
+                return Err(Status::invalid_argument(format!(
+                    "Input amount {} exceeds max order {} for keyset",
+                    u64::from(input.amount),
+                    max_value
+                )));
+            }
+        }
+
+        for output in &outputs {
+            let keyset_info = match db_node::keyset::get_keyset::<String>(&mut conn, &output.keyset_id).await {
+                Ok(info) => info,
+                Err(e) => return Err(Status::internal(e.to_string())),
+            };
+            
+            let max_order_u8 = keyset_info.max_order();
+            let max_value = if max_order_u8 >= 64 {
+                u64::MAX
+            } else {
+                (1u64 << max_order_u8) - 1
+            };
+            
+            if u64::from(output.amount) > max_value {
+                return Err(Status::invalid_argument(format!(
+                    "Output amount {} exceeds max order {} for keyset",
+                    u64::from(output.amount),
+                    max_value
+                )));
+            }
+        }
```

Comment:
This would do the work, but, it add two more iterations and cache reads.
The thing is we are already reading keyset_info for each output inside `check_outputs_allow_multiple_units`, so you can do the check there.
And same this with input, we are looping over those in `process_swap_inputs`, so do your check there